imap_processing/ena_maps/utils/spatial_utils.py

```diff
@@ -0,0 +1,270 @@
+
+
+@typing.no_type_check
+def rewrap_even_spaced_el_az_grid(
+    raveled_values: NDArray,
+    shape: tuple[int] | None = None,
+    extra_axis: bool = False,
+    order: typing.Literal["C"] | typing.Literal["F"] = "F",
+) -> NDArray:
+    """
+    Take an unwrapped (raveled) 1D array and reshapes it into a 2D el/az grid.
+
+    Assumes the following must be true of the original grid:
+    1. Grid was evenly spaced in angular space,
+    2. Grid had the same spacing in both azimuth and elevation.
+    3. Elevation is the 0th axis (and extends a total of 180 degrees),
+    4. Azimuth is the 1st axis (and extends a total of 360 degrees).
+
+    Parameters
+    ----------
+    raveled_values : NDArray
+        1D array of values to be reshaped into a 2D grid.
+    shape : tuple[int], optional
+        The shape of the original grid, if known, by default None.
+        If None, the shape will be inferred from the size of the input array.
+    extra_axis : bool, optional
+        If True, input is a 2D array with latter axis being 'extra', non-spatial axis.
+        This axis (e.g. energy bins) will be preserved in the reshaped grid.
+    order : {'C', 'F'}, optional
+        The order in which to rewrap the values, by default 'F' for Fortran order.
+
+    Returns
+    -------
+    NDArray
+        The reshaped 2D grid of values.
+
+    Raises
+    ------
+    ValueError
+        If the input is not a 1D array or 2D array with an extra axis.
+    """
+    if raveled_values.ndim not in (1, 2) or (
+        raveled_values.ndim == 2 and not extra_axis
+    ):
+        raise ValueError("Input must be a 1D array or 2D array with extra axis.")
+
+    # We can infer the shape if its evenly spaced and 2D
+    if not shape:
+        spacing_deg = 1 / np.sqrt(raveled_values.shape[0] / (360 * 180))
+        shape = (int(180 // spacing_deg), int(360 // spacing_deg))
+
+    if extra_axis:
+        shape = (shape[0], shape[1], raveled_values.shape[1])
+    return raveled_values.reshape(shape, order=order)
+
+
+class AzElSkyGrid:
+    """
+    Representation of a 2D grid of azimuth and elevation angles covering the sky.
+
+    All angles are stored internally in radians, but can be accessed in degrees, by
+    appending "_degrees" to the attribute name. For example, `spacing` in radians
+    can be accessed as `spacing_degrees` in degrees.
+
+    Parameters
+    ----------
+    spacing_deg : float, optional
+        Spacing of the grid in degrees, by default 0.5.
+    centered_azimuth : bool, optional
+        Whether the azimuth grid should be centered around 0 degrees/0 radians,
+        i.e. from -pi to pi radians, by default False.
+        If True, the azimuth grid will be from -pi to pi radians.
+    centered_elevation : bool, optional
+        Whether the elevation grid should be centered around 0 degrees/0 radians,
+        i.e. from -pi/2 to pi/2 radians, by default True.
+        If False, the elevation grid will be from 0 to pi radians.
```

Comment:
Should these be `center_azimuth : float` where a user would input what they want the center to be? The bool is slightly confusing to me.

Reply:
Hmmm, I find that somewhat more confusing than just asking if a user wants their azimuth centered around 0 (-180˚ -> 180˚) or simply incrementing (0 -> 360˚). 

If this is confusing, I'd probably lean towards just making it always have centered elevation and non-centered azimuth. Thoughts?

Reply:
Sounds good, I would say fix it to one choice then and do the coordinates you think you'll be working in. Reduce the number of options for now, and then if you do need to change it later we can add that keyword argument then keeping the default at what you choose.

Reply:
Done!